File: goblinframework-monitor/src/main/java/org/goblinframework/monitor/module/monitor/event/InstructionTranslatorManager.java
package org.goblinframework.monitor.module.monitor.event;

import org.goblinframework.core.exception.GoblinDuplicateException;
import org.goblinframework.core.monitor.Instruction;
import org.goblinframework.core.monitor.InstructionTranslator;
import org.goblinframework.core.util.ServiceInstaller;
import org.jetbrains.annotations.NotNull;

import java.util.IdentityHashMap;
import java.util.Map;

abstract class InstructionTranslatorManager {

  private static final Map<Class<? extends Instruction>, InstructionTranslator> buffer;

  static {
    buffer = new IdentityHashMap<>();
    ServiceInstaller.installedList(InstructionTranslator.class).forEach(e -> {
      @SuppressWarnings("unchecked")
      Class<? extends Instruction> type = (Class<? extends Instruction>) e.type();
      if (buffer.put(type, e) != null) {
        throw new GoblinDuplicateException();
      }
    });
  }

  @NotNull
  static InstructionTranslator translator(@NotNull Instruction instruction) {
    return buffer.getOrDefault(instruction.getClass(), new InstructionTranslator() {
      @NotNull
      @Override
      public Class type() {
        throw new UnsupportedOperationException();
      }

      @NotNull
      @Override
      public String translate(@NotNull Instruction instruction, boolean pretty) {
        return pretty ? instruction.asLongText() : instruction.asShortText();
      }
    });
  }
}

File: goblinframework-monitor/src/main/java/org/goblinframework/monitor/instruction/InstructionTranslatorManager.kt
package org.goblinframework.monitor.instruction

import org.goblinframework.api.annotation.Install
import org.goblinframework.api.annotation.Singleton
import org.goblinframework.api.annotation.ThreadSafe
import org.goblinframework.api.service.GoblinManagedBean
import org.goblinframework.api.service.GoblinManagedObject
import org.goblinframework.core.exception.GoblinDuplicateException
import org.goblinframework.core.module.spi.RegisterInstructionTranslator
import org.goblinframework.api.monitor.Instruction
import org.goblinframework.api.monitor.InstructionTranslator1
import java.util.*
import java.util.concurrent.locks.ReentrantReadWriteLock
import kotlin.concurrent.read
import kotlin.concurrent.write

@Singleton
@ThreadSafe
@GoblinManagedBean(type = "monitor")
class InstructionTranslatorManager private constructor()
  : GoblinManagedObject(), InstructionTranslatorManagerMXBean, RegisterInstructionTranslator {

  companion object {
    @JvmField val INSTANCE = InstructionTranslatorManager()
  }

  private val lock = ReentrantReadWriteLock()
  private val translators = IdentityHashMap<Class<out Instruction>, InstructionTranslator1<out Instruction>>()
  private val defaultTranslator: InstructionTranslator1<Instruction>

  init {
    defaultTranslator = InstructionTranslator1 { instruction, pretty ->
      if (pretty) instruction.asLongText() else instruction.asShortText()
    }
  }

  override fun <E : Instruction> register(type: Class<E>, translator: InstructionTranslator1<E>) {
    lock.write {
      translators.putIfAbsent(type, translator)?.run { throw GoblinDuplicateException() }
    }
  }

  @Suppress("UNCHECKED_CAST")
  fun translator(instruction: Instruction): InstructionTranslator1<Instruction> {
    return lock.read {
      translators[instruction.javaClass]
    } as? InstructionTranslator1<Instruction> ?: defaultTranslator
  }

  @Install
  class Installer : RegisterInstructionTranslator by INSTANCE
}